File: filethumbnails/js/js/scenario5.js
//// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
//// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
//// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//// PARTICULAR PURPOSE.
////
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var page = WinJS.UI.Pages.define("/html/scenario5.html", {
        ready: function (element, options) {
            document.getElementById("group-thumb-start").addEventListener("click", pickFolder, false);
        }
    });

    function pickFolder() {

        // Set API parameters
        var requestedSize = 200,
            thumbnailMode = Windows.Storage.FileProperties.ThumbnailMode.picturesView;

        // Clean output in case of repeat usage
        cleanOutput();

        var openpicker = new Windows.Storage.Pickers.FolderPicker(),
            monthShape = Windows.Storage.Search.CommonFolderQuery.groupByMonth;
        openpicker.fileTypeFilter.replaceAll([".jpg", ".png", ".bmp", ".gif", ".tif"]);
        openpicker.suggestedStartLocation = Windows.Storage.Pickers.PickerLocationId.picturesLibrary;
        openpicker.pickSingleFolderAsync().done(function (folder) {
            if (folder) {
                // Verify the folder is in a library since file groups only work in libraries
                if (folder.isCommonFolderQuerySupported(monthShape)) {
                    // Convert folder to file group and query for items
                    var query = folder.createFolderQuery(monthShape);
                    query.getFoldersAsync().done(function (monthList) {
                        if (monthList && monthList.size > 0) {
                            var firstMonth = monthList.getAt(0);
                            firstMonth.getThumbnailAsync(thumbnailMode, requestedSize).done(
                                function (thumbnail) {
                                    if (thumbnail) {
                                        outputResult(firstMonth, thumbnail, "ThumbnailMode.picturesView", requestedSize, true);
                                    } else {
                                        WinJS.log && WinJS.log(SdkSample.errors.noImages, "sample", "status");
                                    }
                                },
                                function (error) {
                                    WinJS.log && WinJS.log(SdkSample.errors.fail, "sample", "status");
                                }
                            );
                            // Then display the list of files in the file group
                            document.getElementById("group-thumb-outputDetails").innerText = "List of files in the group:";
                            firstMonth.getFilesAsync().done(function (files) {
                                if (files.size > 0) {
                                    files.forEach(function (file) {
                                        outputHeader(file.name);
                                    });
                                }
                            });
                        } else {
                            WinJS.log && WinJS.log(SdkSample.errors.emptyFilegroup, "sample", "status");
                        }
                    });
                } else {
                    WinJS.log && WinJS.log(SdkSample.errors.filegroupLocation, "sample", "status");
                }
            } else {
                WinJS.log && WinJS.log(SdkSample.errors.cancel, "sample", "status");
            }
        });
    }

    function outputHeader(string) {
        var outputElement = document.getElementById("group-thumb-outputDetails");
        var outputGroup = document.createElement("div");

        var headerElement = document.createElement("span");
        headerElement.textContent = string;

        outputGroup.appendChild(headerElement);
        outputElement.appendChild(outputGroup);
    }

    function outputResult(item, thumbnailImage, thumbnailMode, requestedSize) {
        document.getElementById("group-thumb-imageHolder").src = URL.createObjectURL(thumbnailImage, { oneTimeOnly: true });
        // Close the thumbnail stream once the image is loaded
        document.getElementById("group-thumb-imageHolder").onload = function () {
            thumbnailImage.close();
        };
        document.getElementById("group-thumb-modeName").innerText = thumbnailMode;
        document.getElementById("group-thumb-groupName").innerText = "Group used: " + item.name;
        document.getElementById("group-thumb-requestedSize").innerText = "Requested size: " + requestedSize;
        document.getElementById("group-thumb-returnedSize").innerText = "Returned size: " + thumbnailImage.originalWidth + "x" + thumbnailImage.originalHeight;
    }

    function cleanOutput() {
        WinJS.log && WinJS.log("", "sample", "status");
        document.getElementById("group-thumb-imageHolder").src = "/images/placeholder-sdk.png";
        document.getElementById("group-thumb-modeName").innerText = "";
        document.getElementById("group-thumb-groupName").innerText = "";
        document.getElementById("group-thumb-requestedSize").innerText = "";
        document.getElementById("group-thumb-returnedSize").innerText = "";
        document.getElementById("group-thumb-outputDetails").innerText = "";
    }
})();

File: Samples/FileThumbnails/cppwinrt/Scenario5.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario5.h"
#include "Scenario5.g.cpp"
#include "SampleConfiguration.h"
#include <sstream>

using namespace winrt;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::Storage;
using namespace Windows::Storage::FileProperties;
using namespace Windows::Storage::Pickers;
using namespace Windows::Storage::Search;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Navigation;

namespace winrt::SDKTemplate::implementation
{
    Scenario5::Scenario5()
    {
        InitializeComponent();
    }

    void Scenario5::OnNavigatedTo(NavigationEventArgs const&)
    {
        Helpers::ResetOutput(rootPage, ThumbnailImage(), OutputTextBlock(), OutputDetails());
    }

    fire_and_forget Scenario5::GetThumbnailButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        Helpers::ResetOutput(rootPage, ThumbnailImage(), OutputTextBlock(), OutputDetails());

        // Pick a folder
        FolderPicker folderPicker;
        folderPicker.FileTypeFilter().ReplaceAll(FileExtensions::Image());
        StorageFolder folder = co_await folderPicker.PickSingleFolderAsync();

        if (folder != nullptr)
        {
            const CommonFolderQuery monthShape = CommonFolderQuery::GroupByMonth;
            // Verify queries are supported because they are not supported in all picked locations.
            if (folder.IsCommonFolderQuerySupported(monthShape))
            {
                // Convert folder to file group and query for items
                IVectorView<StorageFolder> months = co_await folder.CreateFolderQuery(monthShape).GetFoldersAsync();

                if (months.Size() > 0)
                {
                    const ThumbnailMode thumbnailMode = ThumbnailMode::PicturesView;
                    const uint32_t size = 200;
                    StorageFolder firstMonth = months.GetAt(0);
                    StorageItemThumbnail thumbnail = co_await firstMonth.GetThumbnailAsync(thumbnailMode, size);
                    if (thumbnail != nullptr)
                    {
                        Helpers::DisplayResult(ThumbnailImage(), OutputTextBlock(), thumbnailMode, size, firstMonth, thumbnail, true);

                        // Also display the hierarchy of the file group to better visualize where the thumbnail comes from
                        IVectorView<StorageFile> files = co_await firstMonth.GetFilesAsync();

                        std::wstringstream output;
                        output << std::endl;
                        output << L"List of files in this group:" << std::endl << std::endl;
                        for (StorageFile const& file : files)
                        {
                            output << std::wstring_view{ file.Name() } << std::endl;
                        }
                        OutputDetails().Text(output.str());
                        thumbnail.Close();
                    }
                    else
                    {
                        rootPage.NotifyUser(Errors::NoImages, NotifyType::StatusMessage);
                    }
                }
                else
                {
                    rootPage.NotifyUser(Errors::FileGroupEmpty, NotifyType::StatusMessage);
                }
            }
            else
            {
                rootPage.NotifyUser(Errors::FileGroupLocation, NotifyType::StatusMessage);
            }
        }
        else
        {
            rootPage.NotifyUser(Errors::Cancel, NotifyType::StatusMessage);
        }
    }
}
